Serialize first-class version of functions

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19694 Remove duplicate inlineCallToCode
* **#19603 Serialize first-class version of functions**
* #19371 Trace directly into first-class module form.
* #19334 @torch.jit.script(fn) now is a torch.jit.Function
* #19333 pybind CompilationUnit and Function directly

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

Differential Revision: [D15043730](https://our.internmc.facebook.com/intern/diff/D15043730)

zdevito committed:
Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

suo commented on torch/csrc/jit/passes/python_print.h:
```diff
@@ -12,33 +12,21 @@ struct Method;
-    std::vector<ClassTypePtr>& class_table,
-    bool enforce_importable = false);
-
 TORCH_API void PythonPrint(
```
not for this PR, but I feel like this signature is getting perilously large with a lot of communication going on between the exporter and PythonPrint in the form of various in-out params. We're not enforcing a clean abstraction boundary here.

suo reviewed (approved):
lgtm

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head

zdevito committed:
Update on "Serialize first-class version of functions"

Serialize first-class version of functions

Summary: updates the way we serialize to write out the first-class
version of the function. This allows the removal of hacks in the
import pathway for the module hooks. This also updates the way forks
are serialized such that they are no longer required to be stand alone
functions/methods at the top level. Instead they are serialized as
inner functions and enough compiler infrastructure is present to allow
fork to invoke one of those inner functions on reload

gh-metadata: pytorch pytorch 19603 gh/zdevito/25/head
